misk/src/main/kotlin/misk/hibernate/RealTransacter.kt

```diff
@@ -41,5 +47,62 @@ internal class RealTransacter(
     override fun <T : DbEntity<T>> load(id: Id<T>, type: KClass<T>): T {
       return session.get(type.java, id)
     }
+
+    override fun shards(): List<Shard> {
+      if (config.type == DataSourceType.VITESS) {
+        return doWork { connection ->
+          val shards = ArrayList<Shard>()
+          connection.createStatement().use { statement ->
+            val resultSet = statement.executeQuery("SHOW VITESS_SHARDS")
+            while (resultSet.next()) {
+              shards.add(parseShard(resultSet.getString(1)))
+            }
+          }
+          shards
+        }
+      } else {
+        return SINGLE_SHARD_LIST
+      }
+    }
+
+    private fun parseShard(string: String): Shard {
+      val (keyspace, shard) = string.split('/', limit = 2)
+      return Shard(Keyspace(keyspace), shard)
+    }
+
+    override fun <T> target(shard: Shard, function: () -> T): T {
+      if (config.type == DataSourceType.VITESS) {
+        val previousTarget = doWork {
```

Comment:
what’s the scope of a USE statement? I’m anxious Hikari will give us 3 different physical connections for these 3 calls

Reply:
I don't think it will but I agree that calling the function inside `doWork` is clearer.